node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -921,13 +924,58 @@ public Mutex lock(ApplicationId application, Duration timeout) {
     /** Create a lock which provides exclusive rights to modifying unallocated nodes */
     public Mutex lockUnallocated() { return db.lockInactive(); }
 
-    /** Acquires the appropriate lock for this node */
-    public Mutex lock(Node node) {
+    /** Returns the unallocated/application lock, and the node acquired under that lock. */
+    public Optional<NodeMutex> lockAndGet(Node node) {
+        Node staleNode = node;
+
+        for (int i = 0; i < 4; ++i) {
+            Mutex lockToClose = lock(staleNode);
+            try {
+                Optional<Node> freshNode = getNode(staleNode.hostname(), staleNode.state());
```

Comment:
This is just an optimization, right? I don't think it's necessary at this level, but consider adding a comment explaining why `getNode` is called twice.

Reply:
Yes, done.